### PR TITLE
Final-version Refactoring 

### PR DESCRIPTION
**Pull request by ldr426**
This pull request aims to introduce the final version of the Refactoring for the Echoshader project. 
This PR aims to implements all functionalities with relative control widgets, dynamical box, and linkage to enhance the project's capabilities and user experience.

The PR includes four key files that provide essential functionalities:

- `core.py`: This class acts as a central handler for managing echograms, maps, and statistics. It serves as a superstructure with interfaces for plotting, control widgets, bounds, and links.
- `echogram.py`: This file contains functions related to echograms, including single frequency echograms and tricolor echograms.
- `map.py`: This file houses functions for handling geographical aspects, such as map tiles and the geo-position of ships or stationary platforms.
- `curtain.py`: This file contains functions related to 2.5D curtain.
- `box.py`: This file contains functions related to obtaining and plotting box/polygon stream .
- `hist.py`: This file contains functions related to histogram and statistical tables.
- `utils.py`: This file contains configurations for plottings, such as width settings.

Notes about this PR: https://docs.google.com/document/d/1QL89eLLcnLjhV8FRp4L_V1n52z0EbTwrrceGELBhVes/edit

Features built:
https://github.com/OSOceanAcoustics/echoshader/issues/111
https://github.com/OSOceanAcoustics/echoshader/issues/112
https://github.com/OSOceanAcoustics/echoshader/issues/113
https://github.com/OSOceanAcoustics/echoshader/issues/117
https://github.com/OSOceanAcoustics/echoshader/issues/119
https://github.com/OSOceanAcoustics/echoshader/issues/115
https://github.com/OSOceanAcoustics/echoshader/issues/116
https://github.com/OSOceanAcoustics/echoshader/issues/86

small Issues solved:
https://github.com/OSOceanAcoustics/echoshader/issues/107
https://github.com/OSOceanAcoustics/echoshader/issues/110
https://github.com/OSOceanAcoustics/echoshader/issues/99
https://github.com/OSOceanAcoustics/echoshader/issues/133
https://github.com/OSOceanAcoustics/echoshader/issues/132
https://github.com/OSOceanAcoustics/echoshader/issues/129
https://github.com/OSOceanAcoustics/echoshader/issues/127
https://github.com/OSOceanAcoustics/echoshader/issues/125
https://github.com/OSOceanAcoustics/echoshader/issues/96

### CODE REVIEW

**Comment by valentina-s** [comment]
@ldr426 Could you reorganize the repo to remove the `new_version_3` folder so that it is ready for merging following the standard package structure. 

  |-- echoshader
        |-- old_versions_folders (we can remove them once we settle )
        |-- ...
        |-- core.py
        |-- echogram.py
        |-- xyz.py (other scripts)
        |-- tests (a folder containing test scripts corresponding to each module)
              |-- test_core.py
              |-- test_echogram.py
              |-- test_xyz.py
  |-- notebooks

**Comment by ldr426** [comment]
Hey @valentina-s, Thanks! Also need to update `requirement.txt`  and add env creating command as I told you before:
`mamba create -n echoshader-dev -c pyviz -c conda-forge echopype=0.7.1 hvplot geoviews pyvista ipykernel`
This related to https://github.com/OSOceanAcoustics/echoshader/issues/120

**Comment by valentina-s** [comment]
Also a few other notes:
* I see the code for curtain and hist but I cold not make an example with them. Are they functional? I know above you say only **single frequency echogram**, **tile** and **track** and those worked well, but I just tried the other ones since I saw the code.
* Can you add an example passing an extra parameter through **kwargs.

**Comment by valentina-s** [comment]
> Hey @valentina-s, Thanks! Also need to update `requirement.txt` and add env creating command as I told you before: `mamba create -n echoshader -c pyviz -c conda-forge echopype hvplot geoviews pyvista ipykernel` This related to #120

So if the assumption is that people have created the MVBS_Sv dataset, they would not need echopype any more.

**Comment by ldr426** [comment]
Hey Valentina. I see. The complete version and corresponding jupyter demos will be uploaded this night (3/Aug)
> * I see the code for curtain and hist but I cold not make an example with them. Are they functional? I know above you say only **single frequency echogram**, **tile** and **track** and those worked well, but I just tried the other ones since I saw the code.
> * Can you add an example passing an extra parameter through **kwargs.

File: echoshader/new_version_3/core.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+        self.track_box = None
+
+        self.MVBS_ds_gram_box = self.MVBS_ds
+
+        self.MVBS_ds_track_box = self.MVBS_ds
+
+    def echogram(
+        self,
+        channel: List[str],
```
Comment (valentina-s) [inline echoshader/new_version_3/core.py:151]: Could we make it if they do not provide a channel, then all channels are plotted. In many cases, people may preselect the channels they care about in the `MVBS_ds` so then they do not have to type the channel name strings every time.

File: echoshader/new_version_3/utils.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+gram_opts = {
```
Comment (valentina-s) [inline echoshader/new_version_3/utils.py:1]: I thought some of those can be passed through kwargs?

File: echoshader/new_version_3/core.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+    @param.depends(
+        "Sv_range_slider.value",
+        "color_map.value",
+        "update_echogram_button.value",
+    )
+    def _echogram_plot(self):
+        echograms = single_echogram(
+            self.MVBS_ds,
+            self.channel[0],
```
[the remark this answers is not in the document]
Comment (ldr426) [reply echoshader/new_version_3/core.py:218]: This is not the latest version. Many small issues have been corrected on my local. Please check back when it is done.

**Comment by valentina-s** [comment]
@ldr426 I see that you have done the bidirectional logic. It is great that it worked out! Could you reorganize the folder structure to remove the new_version_3 folder, so that this PR is ready for merging. We should finalize this version before doing other things, those can be done as separate issues and PRs.

**Comment by valentina-s** [comment]
I managed the import and fiddled with several panel configurations of the widgets. Some of the interactive linking did not work for me, although one direction worked. I can test more with the new notebooks. For now just make the current one properly import the package. And change the PR from Draft to ready for review, so I can merge it.

**Review by valentina-s — APPROVED** [review]

**Pull request merged**